### PR TITLE
Issue #5207: add support for LITERAL_CASE and LITERAL_DEFAULT in LeftCurlyCheck

### PR DESCRIPTION
**Pull request by vaano94**
Issue #5207: I added additional case field for recognizing whether we have left curly brace after case expression.

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java [hunk 1]
```diff
@@ -250,15 +250,17 @@ public void testLineBreakAfter() throws Exception {
...
-            "35:33: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 33),
-            "36:21: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 21),
-            "39:29: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 29),
-            "39:34: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 34),
-            "45:37: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 37),
-            "51:12: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 12),
-            "54:5: " + getCheckMessage(MSG_KEY_LINE_PREVIOUS, "{", 5),
-            "56:19: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 19),
-            "66:1: " + getCheckMessage(MSG_KEY_LINE_PREVIOUS, "{", 1),
+            "30:21: " + getCheckMessage(MSG_KEY_LINE_PREVIOUS, "{", 21),
```
Comment (Vladlis) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java]: Such changes in test are hard to validate, please make a separate test with a separate input for new test cases.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java [hunk 2]
```diff
@@ -26,10 +26,18 @@ else if (x < 0) {
...
+                    {
+                        break;
+                    }
+                    case (1 + 2):
+                    {
+                        break;
+                    }
+                    case 0:
+                        break;
+                    default:
```
Comment (Vladlis) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: I think we also need to start validating `default` block in scope of this issue. Please add test cases for `default` too.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java [hunk 3]
```diff
@@ -26,10 +26,18 @@ else if (x < 0) {
                 else { break; }
                 switch (a)
                 {
-                case 0:
-                    break;
-                default:
-                    break;
+                    case 1:
```
Comment (Vladlis) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: Do we have test cases with `{` which do not produce violations. If no please add.

**Comment by Vladlis** [comment]
From Wercker:
```
[INFO] Starting audit...
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:1198:9: '{' at column 9 should be on the previous line. [LeftCurly]
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:1989:9: '{' at column 9 should be on the previous line. [LeftCurly]
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:2064:9: '{' at column 9 should be on the previous line. [LeftCurly]
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:2089:9: '{' at column 9 should be on the previous line. [LeftCurly]
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:2215:9: '{' at column 9 should be on the previous line. [LeftCurly]
[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:2231:9: '{' at column 9 should be on the previous line. [LeftCurly]
Audit done.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

Failed job: 
https://github.com/checkstyle/checkstyle/blob/master/wercker.yml#L41-L59

Please make a PR to pgjdbc with fixes.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java (hunk 2 above)
Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: @Vladlis This makes sense.
Since you are expanding the case of the issue, please update original issue with this too.
Release notes only go by issue title, so if we don't update it users won't know about us including `default` too.

Comment (Vladlis) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: @rnveach done.
There is even a comment I didn't notice at first :)
https://github.com/checkstyle/checkstyle/issues/5207#issuecomment-345484643

**Comment by romani** [comment]
if pgjdbc pause apply of such PR, for more than 3 days, please  create path and apply it like https://github.com/checkstyle/checkstyle/blob/master/wercker.yml#L52 (additional patch).

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java (hunk 2 above)
Comment (pbludov) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: There is one more missing case: LITERAL_NEW.
```
        new Runnable() {
            @Override
            public void run() {/**/}
        };
```

**Comment by vaano94** [comment]
I have put enough lines of code to deal with DEFAULT too , but the method exceeds 50 lines, which throws the following exception: 
NCSS for this method is 53 (max allowed is 50). [JavaNCSS]
Is there any way to suppress such a check?

**Comment by romani** [comment]
from wercker:
`[ERROR] /pipeline/source/pgjdbc/pgjdbc/target/gen-src/org/postgresql/core/v3/QueryExecutorImpl.java:1198:9: '{' at column 9 should be on the previous line. [LeftCurly]`

I resolved this problem (".....gen-src....") recently in master, please make sure your fix is based on our latest master.


> NCSS for this method is 53 (max allowed is 50). [JavaNCSS] 
Is there any way to suppress such a check?

:) no simple way while i am here :). Whole source code follow that strict rule, I doubt this update is so special.
Please think about refactoring. If you see no good way to do this, please ask us for advice. @Vladlis , please assist.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java (hunk 1 above)
Comment (vaano94) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java (hunk 3 above)
Comment (vaano94) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java [hunk 4]
```diff
@@ -162,6 +165,7 @@ public void setIgnoreEnums(boolean ignoreEnums) {
         return CommonUtils.EMPTY_INT_ARRAY;
     }
 
+    // -@cs[JavaNCSS] Suppress NCSS, because of the cases
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: Instead of turning off the check, you need to rewrite the switch.
Turn the biggest block with many LITERAL_ * cases to "default:" and "default:" to simple "case LITERAL_SWITCH:".
In addition, you need to correct the comment near the "default:" because it is outdated and erroneous.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java [hunk 5]
```diff
@@ -38,8 +38,9 @@
  * LITERAL_IF},  {@link TokenTypes#LITERAL_SWITCH LITERAL_SWITCH},  {@link
  * TokenTypes#LITERAL_SYNCHRONIZED LITERAL_SYNCHRONIZED},  {@link
  * TokenTypes#LITERAL_TRY LITERAL_TRY},  {@link TokenTypes#LITERAL_WHILE
- * LITERAL_WHILE},  {@link TokenTypes#STATIC_INIT STATIC_INIT},
- * {@link TokenTypes#LAMBDA LAMBDA}.
+ * LITERAL_WHILE},  {@link TokenTypes#STATIC_INIT STATIC_INIT}, {@link
+ * TokenTypes#LITERAL_CASE LITERAL_CASE}, {@link TokenTypes#LITERAL_DEFAULT
+ * LITERAL_DEFAULT}, {@link TokenTypes#LAMBDA LAMBDA}.
  * </p>
  *
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: The first sentence is really bad. See how it looks in the rendered docs:
http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/blocks/package-summary.html
Please fix it.

Comment (vaano94) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: Could you please explain me what's bad in the first sentence?

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: The first sentence should be short and clear. In this check it should be something like "Checks the placement of left curly braces."

The RightCurlyCheck is a good example.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java (hunk 2 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: @pbludov , please move this out to separate issue, please recheck existing issues.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java (hunk 5 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: I am ok to skip this update(do not change) in scope of this issue, could be done in separate issue.
To keep this PR only on target update.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java [hunk 6]
```diff
@@ -186,18 +190,13 @@ public void visitToken(DetailAST ast) {
...
-            case TokenTypes.LITERAL_WHILE:
-            case TokenTypes.LITERAL_CATCH:
-            case TokenTypes.LITERAL_SYNCHRONIZED:
-            case TokenTypes.LITERAL_FOR:
-            case TokenTypes.LITERAL_TRY:
-            case TokenTypes.LITERAL_FINALLY:
-            case TokenTypes.LITERAL_DO:
-            case TokenTypes.LITERAL_IF:
-            case TokenTypes.STATIC_INIT:
-            case TokenTypes.LAMBDA:
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: I do not expect removal of supported tokens, please explain this update.
they are still present in getAcceptableTokens so this is really weird. 
Regression testing will be required

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java (hunk 4 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: to avoid violation of JavaNCSS, you need to make some case blocks like:
```
startToken = xxxxxxxxxxxx(wwww);
brace = xxxxxxxxxx(zzzz);
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java [hunk 7]
```diff
@@ -26,10 +26,10 @@ else if (x < 0) {
                 else { break; }
                 switch (a)
                 {
-                case 0:
-                    break;
-                default:
-                    break;
+                    case 0:
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: this looks like an unnecessary change. please undo.

Comment (vaano94) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyLineBreakAfter.java]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java (hunk 6 above)
Comment (vaano94) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: returned cases back

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java (hunk 5 above)
Comment (vaano94) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: I changed class doc a little bit. Now it looks similar to the RightCurlyCheck

**Comment by romani** [comment]
from wercker:
`[INFO] --- maven-checkstyle-plugin:2.17:check (default-cli) @ postgresql ---
[INFO] Starting audit...
[ERROR] /pipeline/source/pgjdbc/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java:1198:9: '{' at column 9 should be on the previous line. [LeftCurly] .....`

https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java#L1098

this is not expected violation.

**Comment by vaano94** [comment]
@romani Please note, it's 1198, and not 1098. 
[https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/core/v3/QueryExecutorImpl.java#L1198]

**Comment by romani** [comment]
> Please note, it's 1198, and not 1098.

thanks a lot, violation on this line is expected.

**Comment by Vladlis** [comment]
@romani @rnveach please review again, pgjdbc violations are fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java [hunk 8]
```diff
@@ -295,6 +297,53 @@ private static DetailAST findLastAnnotation(DetailAST modifiers) {
...
+
+    /**
+     * Find brace, considering rules for
+     * {@code TokenTypes.INTERFACE_DEF}, {@code TokenTypes.CLASS_DEF},
+     * {@code TokenTypes.ANNOTATION_DEF}, {@code TokenTypes.ENUM_DEF},
+     * {@code TokenTypes.ENUM_CONSTANT_DEF}.
+     * @param ast {@code DetailAST}.
+     * @return {@code DetailAST} brace itself or null if it is not found.
+     */
+    private static DetailAST findBraceForClassDefinition(DetailAST ast) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: please name it findBraceForTypeDefinition , as you process not only classes.
Please be accurate with names.

Comment (vaano94) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java]: done

**Comment by romani** [comment]
code is ok to merge, lets see what regression show us t make final decision on merge

**Comment by rnveach** [comment]
@vaano94 ping! Do you have problems running regression?

**Comment by romani** [comment]
@Vladlis , do you know status on this PR.
Is it abandoned or on pause ?

**Comment by vaano94** [comment]
@rnveach, @romani Sorry,  I have no idea how to run regression tests. 
Could you please help me explaining how to make them and in what way to represent the result? 
I'll do the task ASAP.

**Comment by rnveach** [comment]
@romani I could not find this so please confirm.

https://vaano94.github.io/findbugs/index.html#A1
https://vaano94.github.io/findbugs/xref/Users/Ivan_Kravchenko2/IdeaProjects/contribution-checkstyle/checkstyle-tester/repositories/findbugs/findbugs/src/java/edu/umd/cs/findbugs/OpcodeStack.java.html#L1448
Shouldn't we only be validating brace in case/default if it spans the entire case and nothing comes before or after it?

**Comment by romani** [comment]
> Shouldn't we only be validating brace in case/default if it spans the entire case and nothing comes before or after it?

yes, we have to.
User can do as much inner blocks as he wants for grouping code or .... .We do not care about inner block in this Check. We should care only about `{` right after `case:`

@vaano94 , sorry for long delay in reply, please do update to avoid violation on https://vaano94.github.io/findbugs/index.html#A1, please add such case to UT to prove it.

**Comment by romani** [comment]
Looks like we lost @vaano94

**Comment by pbludov** [comment]
I'm on it

**Comment by romani** [comment]
redone at #6025
